File: experiments/llm-reasoning-cost/generated_code/cpp/07_word_freq_codex.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>
#include <vector>

std::unordered_map<std::string, int32_t> count_words(const std::string& text) {
    std::unordered_map<std::string, int32_t> counts;
    std::istringstream iss(text);
    std::string word;

    while (iss >> word) {
        std::transform(word.begin(), word.end(), word.begin(),
                       [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
        ++counts[word];
    }

    return counts;
}

void print_top(const std::unordered_map<std::string, int32_t>& counts, int32_t n) {
    std::vector<std::pair<std::string, int32_t>> entries;
    entries.reserve(counts.size());

    for (const auto& kv : counts) {
        entries.push_back(kv);
    }

    std::sort(entries.begin(), entries.end(),
              [](const auto& a, const auto& b) {
                  if (a.second != b.second) return a.second > b.second;
                  return a.first < b.first;
              });

    int32_t limit = std::min<int32_t>(n, static_cast<int32_t>(entries.size()));
    for (int32_t i = 0; i < limit; ++i) {
        std::cout << "  " << entries[i].second << ' ' << entries[i].first << '\n';
    }
}

int main() {
    std::string text = "the cat sat on the mat the cat sat";
    auto counts = count_words(text);
    print_top(counts, 5);
    return 0;
}

File: experiments/llm-reasoning-cost/generated_code/rust/07_word_freq_codex.rs
use std::collections::HashMap;

fn count_words(text: &str) -> HashMap<String, i32> {
    let mut counts: HashMap<String, i32> = HashMap::new();
    for word in text.to_lowercase().split_whitespace() {
        *counts.entry(word.to_string()).or_insert(0) += 1;
    }
    counts
}

fn print_top(counts: &HashMap<String, i32>, n: i32) {
    let mut entries: Vec<(&String, &i32)> = counts.iter().collect();

    entries.sort_by(|a, b| {
        b.1.cmp(a.1)
            .then_with(|| a.0.cmp(b.0))
    });

    for (word, count) in entries.into_iter().take(n.max(0) as usize) {
        println!("  {} {}", count, word);
    }
}

fn main() {
    let text = "the cat sat on the mat the cat sat";
    let counts = count_words(text);
    print_top(&counts, 5);
}